openjdk/mmtkBarrierSetC2.cpp

```diff
@@ -67,12 +67,61 @@ void MMTkBarrierSetC2::expand_allocate(
   // The initial slow comparison is a size check, the comparison
   // we want to do is a BoolTest::gt
   bool always_slow = false;
+  // Check if we need initial_slow_test
   int tv = x->_igvn.find_int_con(initial_slow_test, -1);
   if (tv >= 0) {
     always_slow = (tv == 1);
     initial_slow_test = NULL;
+  }
+
+  // We will do MMTk size check.
+  // The implementation tries not to change the current control flow, and instead evaluate allocation size
+  // with max non-los bytes and combine the result with initial_slow_test. If the size check fails, initial_slow_test
+  // should be evaluated to true, and we jump to the slowpath.
+
+  // The max non-los bytes from MMTk
+  int max_non_los_bytes = (int)get_max_non_los_default_alloc_bytes();
+  // Check if allocation size is constant
+  int const_size = x->_igvn.find_int_con(size_in_bytes, -1);
+  if (const_size >= 0) {
+    // Constant alloc size
+    if (const_size > max_non_los_bytes) {
+      // We know at JIT time that we need to go to slowpath
+      always_slow = true;
+      initial_slow_test = NULL;
+    }
   } else {
-    initial_slow_test = BoolNode::make_predicate(initial_slow_test, &x->_igvn);
+    // Variable alloc size
+
+    // Create a node for the constant and compare with size_in_bytes
+    Node *max_non_los_bytes_node = ConLNode::make((long)max_non_los_bytes);
+    x->transform_later(max_non_los_bytes_node);
+    Node *mmtk_size_cmp = new CmpLNode(size_in_bytes, max_non_los_bytes_node);
+    x->transform_later(mmtk_size_cmp);
+    // Size Check: if size_in_bytes >= max_non_los_bytes
+    Node *mmtk_size_bool = new BoolNode(mmtk_size_cmp, BoolTest::ge);
+    x->transform_later(mmtk_size_bool);
+
+    if (initial_slow_test == NULL) {
+      // If there is no previous initial_slow_test, we use the size check as initial_slow_test
+      initial_slow_test = BoolNode::make_predicate(mmtk_size_bool, &x->_igvn);
+    } else {
+      // If there is existing initial_slow_test, we combine the result by 'or' them together.
```

Comment:
Is it the openjdk internal slow test you mentioned before? Can we ignore it and do the slow test completely ourselves?

Reply:
Yes, that's the one I mentioned before. I don't think we can ignore their test. We jump to the slowpath for larger object size, but they jump to the slowpath for various reasons. 
For example the code [here]( https://github.com/mmtk/openjdk/blob/e1e420b3c4fd627218e45ec977972561c559acec/src/hotspot/share/opto/graphKit.cpp#L3582), they check if a bit is set for the type. But that function takes an argument of `extra_slow_test`, if the bit is not set, they will use the `extra_slow_test`. One example of the `extra_slow_test` is [this](https://github.com/mmtk/openjdk/blob/e1e420b3c4fd627218e45ec977972561c559acec/src/hotspot/share/opto/library_call.cpp#L2864), in which they check whether the class is initialized.

I don't feel confident that we can ignore their test. Also this block is only for variable allocation size which should be a rare case anyway.